components/gitpod-protocol/src/frontend-dashboard-service.ts

```diff
@@ -0,0 +1,123 @@
+        activeHeartbeat(): void;
+        setState(state: SetStateData): void;
+    }
+
+    /**
+     * IServer is the server side which is using in dashboard loading screen
+     */
+    export interface IServer extends IServerOn, IServerSend {
+        auth(): Promise<void>;
+    }
+    interface IServerOn {
+        onSetState: Event<SetStateData>;
+        trackEvent(msg: RemoteTrackMessage): void;
+        activeHeartbeat(): void;
+    }
+    interface IServerSend {
+        sendStatusUpdate(status: Status): void;
+        relocate(url: string): void;
+        setSessionID(sessionID: string): void;
+    }
+
+    export interface Status {
+        workspaceID: string;
+        sessionID?: string;
+        loggedUserId: string;
+
+        instanceId?: string;
+        ideUrl?: string;
+        statusPhase?: WorkspaceInstancePhase;
+
+        workspaceDescription: string;
+        workspaceType: string;
+    }
+
+    export interface SetStateData {
+        ideFrontendFailureCause?: string;
+        desktopIDE?: {
+            clientID: string;
+            link: string;
+            label?: string;
+        };
+    }
+
+    /**
+     * interface for post message that send status update from dashboard to supervisor
+     */
+    export interface StatusUpdateEventData {
+        type: "ide-status-update";
+        status: Status;
+    }
+
+    export interface HeartbeatEventData {
+        type: "ide-heartbeat";
+    }
+
+    export interface TrackEventData {
+        type: "ide-track-event";
+        msg: RemoteTrackMessage;
+    }
+
+    export interface RelocateEventData {
+        type: "ide-relocate";
+        url: string;
+    }
+
+    export interface SetSessionIDEventData {
+        type: "ide-set-session-id";
+        sessionID: string;
+    }
+
+    export interface SetStateEventData {
+        type: "ide-set-state";
+        state: SetStateData;
+    }
+
+    export function isStatusUpdateEventData(obj: any): obj is StatusUpdateEventData {
+        return obj != null && typeof obj === "object" && obj.type === "ide-status-update";
+    }
+
+    export function isHeartbeatEventData(obj: any): obj is HeartbeatEventData {
+        return obj != null && typeof obj === "object" && obj.type === "ide-heartbeat";
+    }
+
+    export function isTrackEventData(obj: any): obj is TrackEventData {
+        return obj != null && typeof obj === "object" && obj.type === "ide-track-event";
+    }
+
+    export function isRelocateEventData(obj: any): obj is RelocateEventData {
+        return obj != null && typeof obj === "object" && obj.type === "ide-relocate";
+    }
+
+    export function isSetSessionIDEventData(obj: any): obj is SetSessionIDEventData {
+        return obj != null && typeof obj === "object" && obj.type === "ide-set-session-id";
+    }
+
+    export function isSetStateEventData(obj: any): obj is SetStateEventData {
+        return obj != null && typeof obj === "object" && obj.type === "ide-set-state";
+    }
```

Comment:
nitpick: I think we might be able to refactor this to be just one function? Would remove a lot of duplication and possibly scale better. We could use a `switch` statement in `frontend-dashboard-service.ts` also for a cleaner implementation.

Reply:
We can do something like this 🤔 

```suggestion
    export function isStatusUpdateEventData(obj: any): obj is StatusUpdateEventData {
        return isType(obj, "ide-status-update");
    }

    export function isHeartbeatEventData(obj: any): obj is HeartbeatEventData {
        return isType(obj, "ide-heartbeat");
    }

    export function isTrackEventData(obj: any): obj is TrackEventData {
        return isType(obj, "ide-track-event");
    }

    export function isRelocateEventData(obj: any): obj is RelocateEventData {
        return isType(obj, "ide-relocate");
    }

    export function isSetSessionIDEventData(obj: any): obj is SetSessionIDEventData {
        return isType(obj, "ide-set-session-id");
    }

    export function isSetStateEventData(obj: any): obj is SetStateEventData {
        return isType(obj, "ide-set-state");
    }

    function isType(obj: any, type: string): boolean {
        return obj != null && typeof obj === "object" && obj.type === type;
    }
```

Reply:
Or you can define union type and let ts derive subtype in place.